[Contextual Security] Update graph filtering to use alerts and events timestamps

alexreal1314 opened the pull request:
## Summary

When investigating security alerts in the graph visualization, we encountered a critical issue where log events were not being fetched because of time range mismatches. Specifically:
Alert Creation Time `alert.@timestamp`: When the alert was detected/created (e.g., October 15th)
Original Event Time `kibana.alert.original_time`: When the underlying event actually occurred (e.g., October 1st)
The original implementation only used the alert's @timestamp ±30 minutes to query both alert and log indices, causing logs that occurred days or weeks before the alert to be excluded from the graph.

**Solution Exploration**
We initially considered implementing a dual time range strategy where:
Alert indices would be queried using @timestamp (alert creation time)
Log indices would be queried using kibana.alert.original_time (original event time)
This approach had several significant drawbacks:
1. Complexity: Required passing two separate time ranges (eventTimeStart/eventTimeEnd in addition to start/end) from client to server
2. Timeline Integration Issues**: When users tried to filter by time range in the Timeline, having two different time ranges created confusion about which range was being applied
3. Maintenance Burden: Multiple time range parameters increased cognitive load and made the codebase harder to understand
4. Schema Complexity: Required additional optional fields in the API schema

**Solution Chosen**
a single unified time range calculated on the client side that encompasses both the alert creation time and the original event time - calculate earliest and latest times based on alert and related event timestamps to make sure both are inside the time range when querying by it.

**Benefits:**
✅ Single time range sent to backend - simpler API
✅ Timeline compatibility - one consistent time range for filtering
✅ Captures all events - spans from earliest to latest timestamp with buffers
✅ Cleaner codebase - no dual time range logic in backend

**Example scenario:**
```
Alert created: Oct 15, 14:30
Original event: Oct 1, 10:30
Unified range: Oct 1, 10:00 - Oct 15, 15:00  ✅ Log found
```

### How to test

1. Run Elasticsearch with `yarn es snapshot --license trial -E reindex.remote.whitelist=keep-long-live-env-ess.es.us-west2.gcp.elastic-cloud.com:443`
2. Run Kibana locally with `yarn start --no-base-path`
3. Install Cloud Asset Discovery (skip agent installation)
4. Re-index entity data with this query
    ```
    POST _reindex?wait_for_completion=true
    {
      "conflicts": "proceed", 
      "source": {
        "remote": {
          "host": "${ES_REMOTE_HOST}", # <-- Set var in Config tab to https://keep-long-live-env-ess.es.us-west2.gcp.elastic-cloud.com:443
          "username": "${ES_REMOTE_USER}", # <-- Set var in Config tab to your user in https://keep-long-live-env-ess.kb.us-west2.gcp.elastic-cloud.com/app/management/security/users i.e. alberto
          "password": "${ES_REMOTE_PASS}" # <-- Set var in Config tab to your password in the same cloud env
        },
        "index": ".entities.v1.latest.security_generic_default",
        "query": {
          "bool": {
            "must": [
              {
                "range": {
                  "@timestamp": {
                    "gte": "now-1d"
                  }
                }
              }
            ]
          }
        }
      },
      "dest": {
        "op_type": "create",
        "index": ".entities.v1.latest.security_generic_default"
      }
    }
    ```
5. execute enrich policy - go to http://<kibana_url>/kbn/app/management/data/index_management/enrich_policies -> execute generic entity policy.
6. Reindex AWS Cloudtrail logs with this query
    ```
    POST _reindex?wait_for_completion=false
    {
      "conflicts": "proceed", 
      "source": {
        "remote": {
          "host": "${ES_REMOTE_HOST}",
          "username": "${ES_REMOTE_USER}",
          "password": "${ES_REMOTE_PASS}"
        },
        "index": "logs-aws.cloudtrail-*",
        "query": {
           "bool": {
             "must": [],
             "filter": [
               {
                 "range": {
                   "@timestamp": {
                     "gte": "now-3h",
                     "lte": "now"
                   }
                 }
               }
             ]
           }
         }
      },
      "dest": {
        "op_type": "create",
        "index": "logs-aws.cloudtrail-reindexed"
      }
    }
    ```
7. Create a new Data View that matches this index pattern `logs-aws.cloudtrail-*`. Give it the exact same name
8. Open the data-view you created and copy the id from the url
<img width="1700" height="906" alt="image" src="https://github.com/user-attachments/assets/c3fbf289-9a86-491d-97bb-2b52daf80e13" />

9. Download and decompress this exported rule: [rules_export (2).ndjson.zip](https://github.com/user-attachments/files/22530457/rules_export.2.ndjson.zip). Replace the value of "data_view_id" with the id you copied in the previous step
<img width="659" height="430" alt="image" src="https://github.com/user-attachments/assets/9632b750-0499-49a0-aa9a-9af5fac93839" />

10. Go to Rules page and click on "Import rule"
11. Go to Alerts page. Set the date range from several days ago until now. You'll see a bunch of alerts.
12. Find an alert whos timestamp is more than 30m the event timestamp one alert.

`alert.@timestamp`:
<img width="1710" height="834" alt="image" src="https://github.com/user-attachments/assets/9745d188-2f28-4dac-bc87-ba06676747c9" />

`kibana.alert.original_time`:
<img width="1701" height="829" alt="image" src="https://github.com/user-attachments/assets/b54ef9ae-f9d9-490a-a8df-d05a6b980caf" />

13. expand the alert -> overview tab -> graph preview -> expand the graph
14. graph should be visible in preview and regular modes.

### Checklist

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

### Screen recording

https://github.com/user-attachments/assets/45a38b98-88e7-4fa9-adc3-2b6396adb94e

alexreal1314 committed:
add dual time range filtering using alerts and events timestamps

apply index-specific filtering in Elasticsearch query

update related unit tests, types and schemas

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/354852)
* Commit: 408b944c0d639dab269d590c7362490e5ca66f6f

### Failed CI Steps
* [Checks](https://buildkite.com/elastic/kibana-pull-request/builds/354852#019a1d55-7584-4ad5-bbfa-0a093d100004)

cc @alexreal1314

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/354852","number":354852,"commit":"408b944c0d639dab269d590c7362490e5ca66f6f"}],"number":354852}
buildkite-pr-comment-->

alexreal1314 committed:
move index patterns to consts and update query

alexreal1314 commented on x-pack/solutions/security/plugins/cloud_security_posture/server/routes/graph/fetch_graph.ts:
```diff
@@ -100,48 +112,87 @@ const buildDslFilter = (
+                    {
+                      bool: {
+                        filter: [
+                          { wildcard: { _index: `*${SECURITY_ALERTS_PARTIAL_IDENTIFIER}*` } },
```
i was thinking of maybe switching to `prefix` syntax, eg.:
```
should: [
                    // Alerts: use alert creation time
                    {
                      bool: {
                        filter: [
                          { prefix: { _index: `${SECURITY_ALERTS_PARTIAL_IDENTIFIER}` } },
                          { range: { '@timestamp': { gte: start, lte: end } } },
                        ],
                      },
                    },
                    // Logs: use original event time
                    {
                      bool: {
                        filter: [
                          { prefix: { _index: LOGS_INDEX_PATTERN } },
                          { range: { '@timestamp': { gte: eventTimeStart, lte: eventTimeEnd } } },
                        ],
                      },
                    },
                  ],
```

but after running profile analzer seems that wildcard syantax is faster in about 2.2x.

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/354883)
* Commit: 63dc43fc4eac18d034f2d6a3bdc9079293a43af7
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-240714)

### Failed CI Steps
* [Jest Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062)
* [Jest Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad82-4088-92a2-cf000e927be3) Jest Tests #<span></span>1 / useGraphPreview should return false when missing actor
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae93-43e2-bea9-8b9db8a1b3e1) Jest Tests #<span></span>1 / useGraphPreview should return false when missing actor
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad81-4f8b-aa28-bf9695b4114f) Jest Tests #<span></span>1 / useGraphPreview should return false when missing event.action
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae92-4676-9e5c-197f6e91a425) Jest Tests #<span></span>1 / useGraphPreview should return false when missing event.action
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad80-481f-a35f-d5ef118719e9) Jest Tests #<span></span>1 / useGraphPreview should return false when missing original_event.id
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae92-4b13-8fcb-568c6bc08934) Jest Tests #<span></span>1 / useGraphPreview should return false when missing original_event.id
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad85-4370-8ea1-0321341e7fc7) Jest Tests #<span></span>1 / useGraphPreview should return false when missing target
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae96-4e96-a931-ef46f947e5bc) Jest Tests #<span></span>1 / useGraphPreview should return false when missing target
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad84-4ce3-b84d-a234deb7b56f) Jest Tests #<span></span>1 / useGraphPreview should return false when timestamp is missing
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae95-4d79-958b-882f3a7e438f) Jest Tests #<span></span>1 / useGraphPreview should return false when timestamp is missing
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad88-4234-a45f-578ad799d978) Jest Tests #<span></span>1 / useGraphPreview should return true when alert has graph preview
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae98-4977-ad99-420960bb6ec0) Jest Tests #<span></span>1 / useGraphPreview should return true when alert has graph preview
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad83-4b69-9add-e03515b452cd) Jest Tests #<span></span>1 / useGraphPreview should return true when alert has graph preview with multiple values
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae94-4873-a38c-1a52efafae4b) Jest Tests #<span></span>1 / useGraphPreview should return true when alert has graph preview with multiple values
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad86-408c-a6aa-456d7c3f8cb2) Jest Tests #<span></span>1 / useGraphPreview should return true when event has graph graph preview
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae98-4b8d-bcf9-6ca3d620b5a1) Jest Tests #<span></span>1 / useGraphPreview should return true when event has graph graph preview
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a20f9-8efb-49ed-988a-b18c6d18b062) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a20f9-8efb-49ed-988a-b18c6d18b062/artifacts/019a212e-ad85-4d85-aca2-8bf763384d89) Jest Tests #<span></span>1 / useGraphPreview should return true when event has graph preview with multiple values
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/354883#019a212e-ef92-41e3-85dd-c2ade398962d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/354883/jobs/019a212e-ef92-41e3-85dd-c2ade398962d/artifacts/019a2156-ae96-4843-bfef-1e513b5f1b36) Jest Tests #<span></span>1 / useGraphPreview should return true when event has graph preview with multiple values

The CI Stats report is too large to be displayed here, check out the CI build annotation for this information.

### History
* :broken_heart: [Build #354852](https://buildkite.com/elastic/kibana-pull-request/builds/354852) failed 408b944c0d639dab269d590c7362490e5ca66f6f

cc @alexreal1314

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/354883","number":354883,"commit":"63dc43fc4eac18d034f2d6a3bdc9079293a43af7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/354852","number":354852,"commit":"408b944c0d639dab269d590c7362490e5ca66f6f"}],"number":354883}
buildkite-pr-comment-->

alexreal1314 committed:
fix graph integrations tests

alexreal1314 committed:
use unified time range for graph visualization based on alert and events timestamps

calculate single time range spanning alert creation time and original event time

update relevant unit tests

alexreal1314 committed:
fix lint issues

elasticmachine commented:
Pinging @elastic/kibana-cloud-security-posture (Team:Cloud Security)

albertoblaz reviewed (approved):
IMO the approach makes sense. I'm not sure if we might produce any unexpected side-effect with this change other than showing more nodes in the graph and make it more noisy. But I think it's a simple way to deal with this issue without introducing too much complexity.

Approving this PR, though we should check with @kfirpeled

PhilippeOberti reviewed (approved):
Left a few comments but approving to not block the PR. Code review only, LGTM for the @elastic/security-threat-hunting-investigations team

alexreal1314 committed:
optimized graph visualization component rendering

alexreal1314 committed:
add FTR test for graph unified time range functionality

validates that graph visualization correctly renders alerts and events
from different time periods

alexreal1314 committed:
Merge branch 'main' into 239244-graph-search-timestamps
